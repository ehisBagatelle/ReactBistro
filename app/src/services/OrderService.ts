import { taxes } from "../../test/fixture";
import { formatDecimal } from "../helper";

export type LineItem = {
  name: string;
  category: string;
  price: number;
};

export type Tax = {
  name: string;
  rate: number;
  appliesToCategories: string[];
};

export type Discount = {
  id: number
  type: "dollar" | "percentage";
  amount: number;
};

export type Bill = {
  totalTaxes: string;
  totalDiscounts: string;
  subTotal: string;
  total: string;
};

export interface IOrderService {
  calculateBill(lineItems: LineItem[], appliedDiscounts: Discount[]): Bill;
}

export class OrderService implements IOrderService {
  taxes: Tax[] = taxes;

  public calculateBill(
    lineItems: LineItem[],
    appliedDiscounts: Discount[]
  ): Bill {
    let subTotal = 0;

    // Calculate pre-tax total
    lineItems.forEach((item) => {
      subTotal += item.price;
    });

    // Calculate total taxes
    const totalTaxes = this.calculateTotalTaxes(lineItems);
    
    // Calculate post-tax total
    const postTaxTotal = subTotal + totalTaxes;

    // Calculate total discounts
    const totalDiscounts = this.calculateTotalDiscounts(
      appliedDiscounts,
      postTaxTotal
    );

    const total = postTaxTotal - totalDiscounts

    return {
      totalTaxes: formatDecimal(totalTaxes),
      totalDiscounts: formatDecimal(totalDiscounts),
      subTotal: formatDecimal(subTotal),
      total: formatDecimal(total)
    };
  }

  private calculateTotalTaxes(lineItems: LineItem[]): number {
    let totalTaxes = 0;

    lineItems.forEach((item) => {
      const applicableTaxes = this.taxes.filter((tax) =>
        tax.appliesToCategories.includes(item.category)
      );

      applicableTaxes.forEach((tax) => {
        totalTaxes += item.price * tax.rate;
      });
    });

    return totalTaxes;
  }

  private calculateTotalDiscounts(
    appliedDiscounts: Discount[],
    postTaxTotal: number
  ): number {
    // The order of applied discounts matter and can change the final output totals
    // We recommend going dollar amount discounts first
    let postDiscountTotal = postTaxTotal;
  
    // Separate dollar amount discounts and percentage discounts
    const dollarAmountDiscounts = appliedDiscounts.filter(
      (discount) => discount.type === "dollar"
    );
  
    const percentageDiscounts = appliedDiscounts.filter(
      (discount) => discount.type === "percentage"
    );
  
    // Apply dollar amount discounts first
    dollarAmountDiscounts.forEach((discount) => {
        postDiscountTotal -= discount.amount;
    });
  
    // Apply percentage discounts after dollar amount discounts
    percentageDiscounts.forEach((discount) => {
        postDiscountTotal -= (discount.amount / 100) * postDiscountTotal;
    });
  
    return postTaxTotal - postDiscountTotal;
  }
}
