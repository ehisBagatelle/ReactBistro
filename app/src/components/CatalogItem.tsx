import React from "react";
import { TouchableOpacity, Text } from "react-native";
import { LineItem } from "../services/OrderService";
import { formatDecimal } from "../helper";
import { styles } from "./Components.styles";

interface Props {
  item: LineItem;
  onPress: () => void;
}

const CatalogItem: React.FC<Props> = ({ item, onPress }) => (
  <TouchableOpacity onPress={onPress} style={styles.item}>
    <Text style={styles.title}>{item.name}</Text>
    <Text style={styles.title}>${formatDecimal(item.price)}</Text>
  </TouchableOpacity>
);

export default CatalogItem;
