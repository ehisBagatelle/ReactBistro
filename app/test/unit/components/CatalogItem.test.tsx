import React from "react";
import renderer from "react-test-renderer";
import { TouchableOpacity, Text } from "react-native";
import CatalogItem from "../../../src/components/CatalogItem";
import { LineItem } from "../../../src/services/OrderService";

const mockOnPress = jest.fn();

const sampleItem: LineItem = {
    name: "Test Item",
    category: "Test Category",
    price: 9.99,
    key: "3003"
};

describe("CatalogItem", () => {
  it("renders correctly with item details", () => {
    const tree = renderer
      .create(<CatalogItem item={sampleItem} onPress={mockOnPress} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("calls onPress when TouchableOpacity is pressed", () => {
    const component = renderer.create(
      <CatalogItem item={sampleItem} onPress={mockOnPress} />
    );
    const touchableOpacity = component.root.findByType(TouchableOpacity);

    touchableOpacity.props.onPress();

    expect(mockOnPress).toHaveBeenCalled();
  });
});
