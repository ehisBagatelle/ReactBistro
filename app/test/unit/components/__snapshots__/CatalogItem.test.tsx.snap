// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogItem renders correctly with item details 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
      "padding": 10,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "fontSize": 18,
        "fontStyle": "normal",
        "fontWeight": "400",
      }
    }
  >
    Test Item
  </Text>
  <Text
    style={
      {
        "fontSize": 18,
        "fontStyle": "normal",
        "fontWeight": "400",
      }
    }
  >
    $
    9.99
  </Text>
</View>
`;
